Improve documentation on saga behaviour when using Sql Server transport and persistence

jpalac opened the pull request:
Part of https://github.com/Particular/CustomerSuccess/issues/4120

ramonsmits commented:
I think we should create a separate page that sits under both SQL Server Transport and SQL Persistence and contains all information about combining these:

Menu:

```
Transports
--SQL Server
  -- SQL Persistence integration
Persistence
--SQL
  -- SQL Server transport integration
```

or maybe shorter in the menu like:

```
Transports
--SQL Server
  -- Persistence integration
Persistence
--SQL
  -- Transport integration
```

Maybe "Sharing transport and persistence" or "Combining "

I think it likely helps to put all of this in a table. I don't think this information should be put in the SQL Server Transport transaction guidance.

SQL Transport<br/>TransactionMode | SQL Persistence | Matching dialect | Outbox | Outcome
:-:|:-:|:-:|:-:|:-:
TransactionScope |  ✅| ✅|✅| Connection sharing via SQLT storage context
AtomicSendsWithReceive |  ✅| ✅|✅| SQLP uses isolated transaction
ReceiveOnly |  ✅| ✅|✅| Connection sharing via SQLT storage context
None |  ✅| ✅|✅| Connection sharing via SQLT storage context

jpalac commented:
@ramonsmits I've made some adjustments - see what you think. To make it less complex I've split it into two sections: with and without outbox, used the sql dialect by default so no need for it in the table, and added a column for where the saga is actually stored so that it's spelled out for the user.

johnsimons reviewed (commented):
Is this ready for a review yet?

jpalac commented:
Go for it

________________________________
From: John Simons ***@***.***>
Sent: Monday, June 10, 2024 6:00:45 PM
To: Particular/docs.particular.net ***@***.***>
Cc: Jo Palac ***@***.***>; Author ***@***.***>
Subject: Re: [Particular/docs.particular.net] Improve documentation on saga behaviour when using Sql Server transport and persistence (PR #6690)


@johnsimons commented on this pull request.

Is this ready for a review yet?

—
Reply to this email directly, view it on GitHub<https://github.com/Particular/docs.particular.net/pull/6690#pullrequestreview-2106959868>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ABA5YJFDGVNCAWOEH4NTOZTZGVMK3AVCNFSM6AAAAABI6BJTGCVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDCMBWHE2TSOBWHA>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

abparticular commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,36 @@
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| Connection sharing via SQLT storage context | Transport Db
```
Note, I'm not sure if &lt;sup&gt; will work on docs site

```suggestion
TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
```

jpalac replied:
@abparticular it seems to work ok - I tried it locally first then pushed the change.

abparticular commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,36 @@
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| Connection sharing via SQLT storage context | Transport Db
+AtomicSendsWithReceive |  ✅| SQLP uses isolated transaction for send and receive | Persistence Db
```
```suggestion
AtomicSendsWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```

abparticular commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,36 @@
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| Connection sharing via SQLT storage context | Transport Db
+AtomicSendsWithReceive |  ✅| SQLP uses isolated transaction for send and receive | Persistence Db
+ReceiveOnly |  ✅| SQLP uses isolated transaction for receive | Persistence Db
```
```suggestion
ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
```

abparticular commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,36 @@
+TransactionScope |  ✅| Connection sharing via SQLT storage context | Transport Db
+AtomicSendsWithReceive |  ✅| SQLP uses isolated transaction for send and receive | Persistence Db
+ReceiveOnly |  ✅| SQLP uses isolated transaction for receive | Persistence Db
+None |  ✅| No transactions | Persistence Db
```
```suggestion
None |  ✅| No transactions | Persistence DB

<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```

abparticular commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,36 @@
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅|  Not supported | N/A
+AtomicSendsWithReceive |  ✅| Not supported | N/A
+ReceiveOnly |  ✅| Connection sharing via SQLT storage context | Persistence Db
```
```suggestion
ReceiveOnly |  ✅| Connection sharing via SQLT storage context | Persistence DB
```

abparticular commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
+AtomicSendsWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```
```suggestion
SendsAtomicWithReceive |  ✅| SQLT uses isolated transaction for send and receive | Transport DB
```

johnsimons commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+ReceiveOnly |  ✅| SQLT uses isolated transaction for receive | Transport DB
+None |  ✅| No transactions | Persistence DB
+
+<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```
Would this only work on Windows OSs?

johnsimons replied:
and if so, should we point it out

johnsimons commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅| SQLT transaction is promoted to distributed transaction | Persistence DB <sup>1</sup>
```
What is `SQLT`, according to google https://www.google.com/search?q=sqlt
I guess my point is that the acronym, is known for something else, I think we need to change it

jpalac replied:
I have changed it to SQL Transport

johnsimons commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+
+#### Explicitly opting out of connection sharing when not using the Outbox
+
+When an endpoint uses SQL Persistence combined with the SQL Server Transport without the [Outbox](/nservicebus/outbox/), the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming of the message that trigge ...
```
```suggestion
When an endpoint uses SQL Persistence combined with the SQL Server Transport with [Outbox](/nservicebus/outbox/) disabled, the persistence uses the connection and transaction context established by the transport when accessing saga data. This behavior ensures *exactly-once* message processing behavior as the state change of the saga is committed atomically while consuming the message that triggered it.
```

johnsimons commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+
+SQL Transport<br/>TransactionMode | SQL Persistence<br/>with Sql dialect | Connection sharing | Saga location
+:-:|:-:|:-:|:-:
+TransactionScope |  ✅|  Not supported | N/A
```
I don't see the purpose of ✅ in the table?

jpalac replied:
I have removed them.

johnsimons commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+
+## Connection behavior
+
+When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is used or not. This influences where the saga data is stored.
```
```suggestion
When combining SQL Server transport and persistence using the Sql dialect, the connection behaves differently based on whether the [Outbox](/nservicebus/outbox/) is enabled or disabled. This influences where the saga data is stored.
```

johnsimons commented on transports/sql/transactions_native_sqltransport_[6,).partial.md:
```diff
@@ -6,3 +6,6 @@ There are two available options within native transaction level:
  * **SendsAtomicWithReceive** - This mode is similar to the `ReceiveOnly`, but transaction is shared with sending operations. That means the message receive operation and any send or publish operations are committed atomically.
+
+> [!WARNING]
+> If using [SQL Server transport](/transports/sql/) and [SQL Persistence with the Microsoft SQL Server dialect](/persistence/sql/dialect-mssql.md) with different connection strings, without [outbox](/nservicebus/outbox/) enabled, in this transaction mode [the persistence uses the connection and transaction context established by the transport when accessing saga data](/persistence/sql/sqlserver-c ...
```
This does not read well, needs improvement.

johnsimons replied:
Maybe first step is to start the sentence with "When using ...."

jpalac replied:
I have reworded this to make it simpler.

johnsimons commented on transports/transactions.md:
```diff
@@ -78,6 +78,10 @@ Some transports support enlisting outgoing operations in the current receive tra
 snippet: TransportTransactionAtomicSendsWithReceive
+
+> [!WARNING]
+> If using [SQL Server transport](/transports/sql/) and [SQL persistence with the Microsoft SQL Server dialect](/persistence/sql/dialect-mssql.md) with different connection strings, without [outbox](/nservicebus/outbox/) enabled, in this transaction mode [the persistence uses the connection and transaction context established by the transport when accessing saga data](/persistence/sql/sqlserver-c ...
```
Same as before

johnsimons commented on persistence/sql/sqlserver-combining-persistence-wth-transport.md:
```diff
@@ -0,0 +1,38 @@
+ReceiveOnly | SQL Transport uses isolated transaction for receive | Transport DB
+None | No transactions | Persistence DB
+
+<sup>1</sup> - Requires .NET Framework, or .NET 8 with `System.Transactions.TransactionManager.ImplicitDistributedTransactions = true;`
```
```suggestion
<sup>1</sup> - Requires .NET Framework, or .NET 8 [`ImplicitDistributedTransactions`](https://learn.microsoft.com/en-us/dotnet/api/system.transactions.transactionmanager.implicitdistributedtransactions?view=net-8.0) set to `true`.
```
